File: leetcode/30-substring-with-concatenation-of-all-words/sliding-window.js
/**
 * @param {string} s
 * @param {string[]} words
 * @return {number[]}
 * @refer [https://leetcode-cn.com/problems/substring-with-concatenation-of-all-words/solution/dong-hua-yan-shi-30-chuan-lian-suo-you-dan-ci-de-z/]
 */
const findSubstring = function(s, words) {
  if (!s || s.length === 0 || !words || words.length === 0) {
    return []
  }

  const map = {}
  const res = []

  for (let w of words) {
    if (!(w in map)) {
      map[w] = 1

    } else {
      map[w] += 1
    }
  }

  const len1stWord = words[0].length
  const lenAllwords = words.length * len1stWord

  for (let i = 0; i < s.length - lenAllwords + 1; i++) {
    const tmpStr = s.slice(i, i + lenAllwords)
    const tmpMap = { ...map }

    for (let j = 0; j < tmpStr.length; j += len1stWord) {
      const key = tmpStr.slice(j, j + len1stWord)

      if (key in tmpMap) {
        tmpMap[key] -= 1

        if (tmpMap[key] === 0) {
          delete tmpMap[key]
        }

      } else {
        break
      }
    }

    if (Object.keys(tmpMap).length === 0) {
      res.push(i)
    }
  }

  return res
}


const s = 'barfoothefoobarman'
const words = ['foo', 'bar']
console.log(findSubstring(s, words))

File: leetcode/30-substring-with-concatenation-of-all-words/sliding_window.py
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        if not s or len(s) == 0 or not words or len(words) == 0:
            return []

        words_map,res = dict(),[]
        for i in words:
            if i not in words_map:
                words_map[i] = 1
            else:
                words_map[i] += 1

        one_word_size = len(words[0])
        all_words_size = len(words) * one_word_size

        for i in range(len(s) - all_words_size + 1):
            tmp_str,d = s[i:i+all_words_size],dict(words_map)

            for j in range(0,len(tmp_str),one_word_size):
                key = tmp_str[j:j+one_word_size]
                if key in d:
                    d[key] -= 1
                    if d[key] == 0:
                        del d[key]
                else:
                    break
            if not d:
                res.append(i)

        return res
